Dt docs for 2.3 part 3, and some test cases as well

mbolivar-nordic opened the pull request:
Along with https://github.com/zephyrproject-rtos/zephyr/pull/25633 and https://github.com/zephyrproject-rtos/zephyr/pull/25470 (already merged), this PR:

Fixes: #25469

By:

- writing a migration guide for moving to the new API
- checking if zephyr v2.2 restriction in https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bindings-file-syntax is still an issue (formerly, only 2 levels of child bindings were permitted, but now we can apparently keep recursing to our heart's delight)

I wanted to make sure the migration guide doesn't regress and the bindings restriction is indeed lifted in perpetuity, so I added test cases as well. I did have to touch gen_legacy_defines.py to make those tests cases not throw warnings about deprecated macros.

With this PR, I also give my sign-off to the DT release notes for #22637.

mbolivar-nordic committed:
tests: devicetree: test great-grandchild bindings

At some point, "child-binding:" apparently only worked up to 2 levels
deep. That's not the case anymore, but add a regression test to make
sure that doesn't break. 3 levels deep ought to be enough for anyone.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: devicetree: update binding-template.yaml

Remove some deprecated or obsolete functionality:

- the 'title:' key is now deprecated
- more than 2 levels of child-binding are now supported

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
tests: move lib/devicetree to lib/devicetree/api

This is preparation for an additional test suite specifically for the
legacy API which will be added next.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: add migration guide documentation and tests

Add test cases that verify various bits and pieces of the legacy
devicetree macros match the new APIs.

Writing these test cases without giving rise to deprecated macro
warnings which might break people's CI if they build with -Werror
requires turning off the __WARN() generation in
devicetree_legacy_unfixed.h. The entire file is deprecated at this
point and must be explicitly enabled with an opt-in Kconfig option, so
there isn't any harm in doing this.

Nevertheless, take a minimally invasive approach to avoiding __WARN()
generation in gen_legacy_defines.py, to avoid the possibility of
breakage. This code is basically frozen anyway, so hacks like this
won't cause maintainability problems since it isn't being actively
maintained.

Use the new tests as fodder for a migration guide from the old API in
the documentation.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@hakonfam you originally proposed the migration guide, so any feedback on this effort would be welcome. Thanks!

pabigot commented on samples/hello_world/app.overlay:
```diff
@@ -0,0 +1,6 @@
+/ {
```
Why is this here?

galak replied:
ditto

mbolivar-nordic replied:
sorry about that

galak reviewed (changes requested):
Looks good, beyond addition of samples/hello_world/app.overlay

mbolivar-nordic committed:
doc: release notes update for devicetree

I believe the devicetree documentation for the release is in good
enough shape now. Add a few more links to the release notes to provide
users with more hints for adapting to the new API.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak reviewed (approved):
Not sure if you can add a license/copyright header to migration-example.inc to make scancode happy.

mbolivar-nordic commented:
> Not sure if you can add a license/copyright header to migration-example.inc to make scancode happy.

I tried that with the C file but I couldn't get the presentation to come out right using `:start-after:`; there was an extra newlines at the start of the file. I'm just going to inline the content even though that's not as nice.
